### PR TITLE
[Master] Decouple OneDNN data structures in MXNet C++ API

### PR DESCRIPTION
**Pull request by mozga-intel**
## Description ##
NDArry file has been modified, there are a few changes:

OneDNN header was moved into *cc file
OneDNN object is created on the fly: static_cast is needed
Decouple oneDNN data structures in MXNet C++ API #20196

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [all]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [unix-cpu, windows-gpu]

**Comment by MateuszOzga** [comment]
@mxnet-bot run ci [centos-gpu , unix-cpu ]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [centos-gpu , unix-cpu ]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [unix-cpu]

File: src/ndarray/ndarray.cc [hunk 1]
```diff
@@ -37,7 +37,9 @@
 #include "../operator/tensor/matrix_op-inl.h"
 #include "../profiler/storage_profiler.h"
 #include "./ndarray_function.h"
-
+#if MXNET_USE_MKLDNN == 1
+#include <mkldnn.hpp>
```
Comment (PawelGlomski-Intel) [inline src/ndarray/ndarray.cc:41]:
```suggestion
#if MXNET_USE_ONEDNN == 1
#include <dnnl.hpp>
```

Comment (mozga-intel) [reply src/ndarray/ndarray.cc]: Good catch!

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [centos-gpu]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [centos-cpu]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [centos-cpu, unix-cpu
]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [windows-cpu, unix-cpu]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [centos-cpu ]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [unix-cpu]

File: include/mxnet/ndarray.h [hunk 2]
```diff
@@ -766,36 +763,36 @@ class NDArray {
   /*
    * This function returns dnnl::memory with the default primitive_desc.
    */
-  const dnnl::memory* GetDNNLData() const;
+  const void* GetDNNLData() const;
```
Comment (anko-intel) [inline include/mxnet/ndarray.h:766]: I agree with @PawelGlomski-Intel  that we can use _dnnl::memory*_ here and below:
```suggestion
  const dnnl::memory* GetDNNLData() const;
```
with forward declaration near line 77
```
namespace dnnl{
  struct memory;
}
```
This way we can avoid most of the static_cast in this change.

File: include/mxnet/ndarray.h [hunk 3]
```diff
@@ -47,6 +44,11 @@
 #error "cxx11 was required for ndarray module"
 #endif
 
+namespace dnnl {
+struct memory;
+struct engine;
```
Comment (anko-intel) [inline include/mxnet/ndarray.h:49]: engine is not used in this file. So why it is forward delared here?

File: include/mxnet/ndarray.h [hunk 4]
```diff
@@ -771,31 +773,31 @@ class NDArray {
...
    * the given primitive_desc.
    */
-  const dnnl::memory* GetDNNLDataReorder(const dnnl::memory::desc& md) const;
+  const dnnl::memory* GetDNNLDataReorder(const void* md) const;
 
   /*
    * This function copies data from dnnl memory.
    */
-  void CopyFrom(const dnnl::memory& mem);
+  void CopyFrom(const void* mem);
```
Comment (anko-intel) [inline include/mxnet/ndarray.h:787]:
```suggestion
  void CopyFrom(const dnnl::memory* mem);
```
I guess using declared above dnnl::memory should make oneDNN implementation not contaminated with many static_cast later (?)

**Review by anko-intel — APPROVED** [review]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [unix-cpu, windows-gpu]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [unix-gpu, website]

**Comment by mozga-intel** [comment]
@mxnet-bot run ci [unix-gpu, website]

**Review by bartekkuncer — APPROVED** [review]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu]

**Pull request merged**